Vary header cache

FSchumacher opened the pull request:
First try on an implementation to support vary header as asked for in
https://bz.apache.org/bugzilla/show_bug.cgi?id=61176

Do not merge.

I haven't written tests yet, as I am not sure, if the code is going in the right direction.

The idea at the moment is to store two entries in the cache, when a vary header is found in the response headers. One entry for the original url with the name of the vary header set and one entry for the url+var-header-name+vary-header-value (of the request).

The code makes it clearer, that the two http samplers with slightly different implementation details make it harder to implement this. For example both samplers have different notions of how the headers are implemented.

FSchumacher committed:
Let the logging facade decide, whether we are in the correct logging mode.

We are not not doing any real work here and it simplifies our code.

FSchumacher committed:
Use log string templates instead of string concatenation.

This should make the code a bit more readable.

FSchumacher committed:
Extract code into local private methods.

Extract the various methods for calculation of the expiration date into
smaller methods in an attempt to make code more readable.

FSchumacher committed:
Rename local variable to comply with java naming conventions.

FSchumacher committed:
Reformat to make it easier to read.

FSchumacher committed:
Use vary header information for caching.

undera reviewed (commented):
First of all - thanks for this initiative. Also I appreciate the refactorings made, project needs them a lot.

One thing that I see is not sufficient is lack of multi-header handling for Vary. RFC allows that and I saw real servers working like that (https://tools.ietf.org/html/rfc7231#section-7.1.4). 
IMO implementation has to use all of headers listed in Vary as part of cache entry key.

And yes, we need unit tests to see how it all will behave as integral.

vlsi reviewed (changes requested):
I'm not sure on the functional side of the feature, however there is a couple of performance-wise issues.

FSchumacher committed:
Guard debug log message with a computation, to be called in debug mode, only

FSchumacher committed:
Remove unused imports.

FSchumacher committed:
A vary header with '*' as value means, that the page should not be cached.

FSchumacher commented on src/protocol/http/org/apache/jmeter/protocol/http/control/CacheManager.java:
```diff
@@ -127,19 +163,28 @@ public Date getExpires() {
+            return null;
+        }
+        for (String headerLine: reqHeaders.split("\n")) {
+            if (headerLine.startsWith(headerName + ": ")) {
```
I am thinking of splitting the `headerLine`, so that I can easier compare the `headerName` ignoring the case.

And as undera pointed out, a vary header can have multiple values (separated by comma), so this logic will have to be changed anyway.

FSchumacher commented on src/protocol/http/org/apache/jmeter/protocol/http/control/CacheManager.java:
```diff
@@ -152,22 +197,18 @@ private boolean hasVaryHeader(URLConnection conn) {
      *            result to decide if result is cacheable
      */
     public void saveDetails(HttpResponse method, HTTPSampleResult res) {
```
This is _old_ code and a result of the different implementations of the http samplers. What refactoring do you have in mind?

FSchumacher committed:
First try to support multiple vary headers.

This depends on HashMap#toString to give consistent results, which is
probably not true.

FSchumacher committed:
Don't loose expire date on calculation

pmouawad commented:
Thanks Felix for PR.
I reviewed it, it looks good to me.
Maybe since we're making API modifications we should start thinking about:
- https://bz.apache.org/bugzilla/show_bug.cgi?id=53540

Thanks

FSchumacher committed:
Tests for vary header are running, except those for multiple vary headers.

FSchumacher committed:
Tests for multi value vary header pass.

undera commented:
Great progress, Felix!
Now we need to test it against real websites with Vary header, right?

FSchumacher commented:
@undera there are still a few tests missing, mainly the ones for vary headers for multiple and different headers. But you are welcome to give this some testing on real sites.

@pmouawad I think that the caching of content/partial content is a different beast and might be looked at after this pr.

FSchumacher committed:
Correct tests for multiple vary headers

FSchumacher committed:
Remove mentioning of the not-caching of Vary headers.

FSchumacher commented:
I think this feature can be tested now.

FSchumacher commented on src/protocol/http/org/apache/jmeter/protocol/http/sampler/HTTPJavaImpl.java:
```diff
@@ -626,6 +627,19 @@ protected HTTPSampleResult sample(URL url, String method, boolean areFollowingRe
         }
     }
 
+    private Header[] getHeaders(HeaderManager headerManager) {
```
I thought of extracting it into a utility class.

pmouawad commented:
Great @FSchumacher  !
Thanks for your work !

pmouawad commented:
Hi Felix,
Looks good to me.

I'd just change a bit in HTTPJavaImpl#getHeaders(HeaderManager headerManager) to:
`    private Header[] getHeaders(HeaderManager headerManager) {
        if (headerManager != null) {
            final CollectionProperty headers = headerManager.getHeaders();
            if (headers != null) {
                List<Header> allHeaders = new ArrayList<>(headers.size());
                for (final JMeterProperty jMeterProperty : headers) {
                    allHeaders.add((Header) jMeterProperty.getObjectValue());
                }
                return allHeaders.toArray(new Header[allHeaders.size()]);
            }
        }
        return new Header[0];
    }`

pmouawad commented:
Hi Felix,
I suggest you commit the PR so that we can update and test code more easily.

Thanks
Regards
